src/compiler/checker.ts

```diff
@@ -8972,8 +8972,7 @@ namespace ts {
         }
 
         function getConstraintDeclaration(type: TypeParameter) {
-            const decl = type.symbol && getDeclarationOfKind<TypeParameterDeclaration>(type.symbol, SyntaxKind.TypeParameter);
-            return decl && getEffectiveConstraintOfTypeParameter(decl);
+            return mapDefined(filter(type.symbol && type.symbol.declarations, isTypeParameterDeclaration), getEffectiveConstraintOfTypeParameter)[0];

```

Comment:
`mapDefined` will return an array and indexing array will produce non-null return type. But in runtime constraintDeclaration can be undefined. Is this desirable?

Reply:
Yeah, that was intentional - I should probably annotate the function to reflect that, so it's checked at callsites.